src/setup_jest.ts

```diff
@@ -57,7 +59,16 @@ const logAndThrow = loggerFn => {
       const err = new Error(message)
       // Skip this frame in the stack to point to the actual log call-site
       Error.captureStackTrace(err, imp)
-      throw err
+
+      // FIXME:
+      // Because we're throwing non-errors, its stopping execution
+      // and failing our tests when library code warns in certain ways. This
+      // creates problems for invariants and other strict checks in
+      // styled-components 4. With the below ccommented out we still get good
+      // logs.
+
+      // throw err
+      // console.log(err)
```

Comment:
This was failing from warnings thrown in lib code (for example, "ref can't be passed to SFC", etc.)

Reply:
@alloy we can return to this

Reply:
I feel strongly that this should be reverted. Warn and error logs will end up getting ignored. Usually it starts with a benign log, which then starts turning the logs red, making the next legit log unnoticeable in the noisy output. Legit logs that we can’t get rid of should selectively be silenced by stubbing `console.*`.

I got here because I noticed the log that I think you’re referring to, why can’t that be fixed?

```
    console.error src/setup_jest.ts:55
      Warning: Stateless function components cannot be given refs. Attempts to acces
s this ref will fail.

      Check the render method of `Autowhatever`.
          in Input (created by Autowhatever)
```

Reply:
This was leading to hundreds of false positives errors across our codebase. Why are we erroring out and failing tests on `Warning:`s?

Reply:
Because over time we were amassing untreated warnings, which start masking more serious warnings, etc etc. Warnings may definitely be ignored for valid reasons, but one should take explicit action to do so. Treating warnings as errors in dev/test mode is a practice employed to make this process explicit.

Reply:
> If you feel there's a way to properly balance things then I'm ok with turning it back on

In JS and Jest it’s fairly easy to do a `console.warn = jest.fn()` before the test that triggers the unavoidable warning (just make sure to reset it in the after filter).

> but this is creating ambiguous failures

Can you elaborate in what sense they’re ambiguous?

Reply:
I understand the reasoning, but again there were warnings, often in lib code, leading to test failures that were not actually failures. If you feel there's a way to properly balance things (and have the time to fix all of the failures) then I'm ok with turning this back on. In the instance of this styled-components 4 upgrade, the balance didn't seem reasonable at the time, hence the large note / comment about returning. This upgrade required a lot of coordination across multiple repos and was a long time coming and so there was an explicit trade off. I don't see any real harm here.

Reply:
See https://reactjs.org/docs/strict-mode.html for more info

Reply:
There appears to be 5 test failures when I disable it, all of which I _think_ are due to the ref passing to a SFC. That hardly seems unwieldy.

Are you perhaps thinking of times when you’re performing large-scale changes (such as a SC upgrade) and it getting in the way of working your way through the various upgrade issues? Because I can totally relate to that, but that’s a case where you can temporarily turn it off until you finish the upgrade.

Reply:
I'm not sure what changed; there was at one point in time many failures. This upgrade spanned two different attempts over a couple months. But anyways, I left the note in the code and tagged you via @ above to review this as it was an obvious change to a very explicit decision, and now you're reviewing, and this is all very by the book. Please revert if you see this as being something easy to fix.  I'm not arguing that it should be any which way, only that it should have been returned to after a coordinated upgrade, which is happening now.

Reply:
I’ll PR a fix for the current failure and also have a think about how to improve the failure output so that it’s easily understood what to do 👍 

> I'm not sure what changed; there was at one point in time many failures.

Perhaps the upstream lib that was throwing the warnings fixed its issue? Hard to say after the fact 🤷‍♂️

Reply:
https://github.com/artsy/reaction/pull/2035